### ISSUE TITLE
A data transmission and processing system for traps and puzzles

### ISSUE THREAD

**Issue by IntegratedQuantum** [PixelGuys/Cubyz#122]
## Requirements
- Data transmission, this could for example be a wire(or wireless) connection from a lever to a door.
- Data validation, this could be a bunch of logic gates, analog circuitry or actual code.
- Measurement devices(levers, pressure sensors, buttons)
- Output devices (doors, secret doors, lights, arrow traps)
- Time utilities (delay, clock, timer)

### Additional ideas:
- toggleable blocks: they are only visible and solid if they get a signal. This would be an easy alternative for piston doors and it would allow making time-limited parkours in dungeons

### Redstone
Pros:
- It doesn't require additionaly GUI overlays
- It is observable in action
- It can be fun to try to figure out how to integrate redstone into a confined space

Cons:
- It is expensive for the client: The mesh needs to update every time the signal changes
- It is expensive for the server: A redstone tick needs to update a lot of blocks
- It has high latency: Redstone ticks are rare and complex calculations require many ticks
- It can clutter your buildings if you don't know all the tricks, like using slabs to transmit a signal upwards
- It would be quite complicated for procedural dungeon generation to generate traps without relying on pre-built structures.
### Integrated Circuits and Wireless Transmission
Essentially you could attach a circuit board to each block, like a lever or a door.
Each circuit board is a square grid, with inputs(e.g. lever state) and outputs(e.g. door state).
The player can place circuit elements(wires, bridges, logic gates) inside a special GUI. These actions would consume items from the player's inventory.
There are wireless receiver and transmitter components. These could be paired up using the same oscillator crystals. Additionally there should be pair up a sender with a receiver to avoid limiting the system too much.

Pros:
- Efficient: no mesh updates, boards could be compiled to simple code and lookup tables.
- No extra space required
- Easier to debug, each circuit could be tested in the GUI.
- You can hack the puzzle if you brute-force the frequency crystal of the door. This fact could be used to make new puzzles as well.

Cons:
- Less immersive than redstone
- Connections are invisible(maybe a tool could make them visible?)
- How to introduce it to the player?
- Is it as fun to use as redstone?

### String
A string has two states: under tension and loose.
Two strings attached to the same endpoint can be used as an OR gate.
A NOT gate can be created using a small weight. The weight tensions the output string, unless the input string lifts it up.

Pros:
- String logic could be built in real life
- It doesn't require additionaly GUI overlays
- It is observable in action, by changing the color of strings that are under tension
- Strings can believably go through walls, so it shares the advantages of the wireless approach
- There might be some fun interactions when players can break strings inside of dungeons
- Doesn't need mesh updates, except from updating the string colors

Cons:
- Collisions could be a problem: Without collisions it would be less immersive, but collisions on the other hand would be expensive, and would tension the strings which can have unforeseen consequences, maybe even make circuits unusable.
- It would still be somewhat expensive for the server, but maybe less expensive than redstone.
- It can absolutely clutter your buildings, maybe multi-colored strings could help
- How does the string interact with ambient light?
- Is it as fun to use as redstone?

### Slime Mold (visually based on Physarum polycephalum)
The mold has two states: idle/sleeping and eating, when eating the texture of the entire organism is animated as nutrients are transferred throughout it.
When eating it can exert some force on objects, which allows it to e.g. open a door or push away food from another instance (→not gate).
Overall this makes it act kind of like redstone, except it can stick on wall and ceiling, and it's alive (be careful not to give it too much food at once, or your circuit breaks).

Pros:
- same as redstone
- more immersive than redstone, it fits better into the world of Cubyz
- placing on walls
- being alive would allow other fun interactions (be careful where you place it, or it will eat it and expand)

Cons:
- expensive (same as redstone)

**Comment by Argmaster** [comment]
I would take into consideration a system similar to Create Mod mechanics with cogs and belts, It had a great success in Minecraft world.

**Comment by IntegratedQuantum** [comment]
That's not a direction I want to go in. I don't want Cubyz to turn into an automation game.
I just want some simple logic for traps and puzzles in dungeons.

**Comment by codemob-dev** [comment]
Maybe something like portal's wire system? Have no length limit for connections and instead cache all points where the wire connects to a powerable block.

**Comment by ikabod-kee** [comment]
I suggest that you guys check out Allumeria's wiring system. It's very smart. Basically you can bind an activator (such as a button or pressure pad) to an activatee. No wires needed.

**Comment by IntegratedQuantum** [comment]
Does it have logic though? E.g. can you make 2 switches toggle the same light source?

Also I cannot check it out, because it doesn't work on linux.

**Comment by ikabod-kee** [comment]
Yes, it's a simple A and B connection. Basically wires that are invisible.

**Comment by codemob-dev** [comment]
Portal has OR gates, we could always make AND gates and XOR gates as well. It's a very simple system where you can attach wires to flat surfaces and they glow when powered. The system isn't accessible outside of the level editor in portal, but the basic mechanics would work in cubyz

**Comment by IntegratedQuantum** [comment]
I would like to mention that we should also choose based on what fits the world and the lore.
I think neither "WLAN cables" nor glowing Portal wires really fit the world we have built so far, which so far is neither hi-technological nor magical, and honestly I think this is the right place to be.

**Comment by codemob-dev** [comment]
Yeah, overall I think string would make a good option. Having actual pulleys to move doors instead of pistons could be cool. Hydraulics are also an option, potentially in the form of pipes embeddable in blocks filled with water or a different liquid.

**Comment by careeoki** [comment]
Pulleys could be pretty unique. If we have climbable ropes you could extend or retract them on a pulley to keep others out of your treehouse or something

**Comment by codemob-dev** [comment]
That would be fun! I don't think I've seen a game that does that before, besides the create mod for Minecraft to some extent.

**Comment by ikabod-kee** [comment]
> That would be fun! I don't think I've seen a game that does that before, besides the create mod for Minecraft to some extent.

Actually, Better Than Wolves did this before Create! In fact it was probably inspired by BTW.

Anyways pulleys are fun (smile, smile)

**Comment by codemob-dev** [comment]
> > That would be fun! I don't think I've seen a game that does that before, besides the create mod for Minecraft to some extent.
> 
> Actually, Better Than Wolves did this before Create! In fact it was probably inspired by BTW.
> 
> Anyways pulleys are fun (smile, smile)

True, I've never gotten very far in BTW so I forgot about it lol

**Comment by ikabod-kee** [comment]
BTW has peak design honestly, it's a big inspiration for me. What's fun about it is that it starts off as a survival game, then slowly transitions into automation. Takes a lot of genuine hard work to get to that point.

**Comment by codemob-dev** [comment]
Yeah, it was really good for its time.

**Comment by IntegratedQuantum** [comment]
I have expanded the list with slime mold, which is my new favorite.

**Comment by careeoki** [comment]
I really really like that, it feels very Cubyz. I'm curious how it would look in action

**Comment by ikabod-kee** [comment]
In action, it would go glorp glorp

**Comment by AproposRobin** [comment]
What about a hybrid approach, Redstone fails in a few categories because of the way it's built your slime mold doesn't have to. As a plus honestly the slime mold could be cool, could make little wooden plant stakes, but built for slime that encourages the slime to grow in a direction. The mold could have the affectation of being electrical(Think electric eels), could hurt the player if you wanted making the circuit open or perhaps adding shorting mechanics(Might be a little feature creep-ish but hey) 

In this instance I will outline a few definitions real quick.
Source - Origin block/item of power
Transmission - Connecting block/item
Trigger - Recipient of power 

Let's look at the two biggest problems plaguing Minecraft's redstone.

- Recursion.
- Block Updates

Recursion itself I think is infinitely easier to solve than what Minecraft has, it relies on the ideal that each dust updates it's power level based on it's neighbors in a strong armed strength check. It does this on tick as well and completely fails to use a network chain.

How we can solve it.
Instead of having recursion happening on the individual instance we make both Source and Trigger owners, with them taking on a master/slave role when connected. When the source propagates it goes from the source out, each level of the Transmision Network receives one update from the source being it's Strength, this is now a child of the Source or Trigger if a source is not attached. If a new Transmision is connected it takes from the current strength to determine it's MaxPossibleStrength and CurrentStrength.

So for i=Source -=Dead Transmission o=Trigger  x(Where x is a value between 0 and 8)=LiveTransmission

Lets look at a full sized full power circuit 

i1234567o
Each distance has a value that's held and retained by it's owner only, it will only update that instance if A it receives power from null or B it's at 0 litteraly. We can slow down the updates intentionally to get cool emergent behavior.

Now lets say we broke the 3 block in slime mold so it looked like this
i12 4567o
On a set timer event(Triggered via an update call from the destroyed block)each tick would slowly dwindle the strength I.E. the Transmission is dying because it isn't connected to it's nutrients i.e. the source. Which would continue until death, deactivating the Trigger. You could also reverse this order for program sake and make it easier to track. 

i12-3456o ->Trigger is notified to take on a master role(Because the trigger is not a source it cannot retain the power so deprecation begins one block at a time all of this is handled only on the trigger though each block is only concerned with on/off state)(Could get really fancy here and utilize bitpack data so 1 byte equals one network chain)
i12 2345o 
i12 1234o
i12 x123o
i12 xx12o
i12 xxx1o
i12 xxxxo ->Trigger loses power

Block updates, can be tougher. I haven't dug to much into your individual updates but this would have to happen as a do once procedure.

**Comment by ikabod-kee** [comment]
<img width="285" height="260" alt="Image" src="https://github.com/user-attachments/assets/1e0b9593-201e-4a8f-8ca0-77c86a9d26f2" />

<img width="385" height="382" alt="Image" src="https://github.com/user-attachments/assets/41d5dfd9-7cf6-4656-a7a9-95efad61143e" />

**Comment by OneAvargeCoder193** [comment]
I don't think the mold itself should push blocks, rather it be an external block like pistons in m word game

**Comment by RanPix** [comment]
what if instead pushing blocks, mold would only pull blocks?

**Comment by OneAvargeCoder193** [comment]
The problem with mold itself pushing and pulling blocks is that it becomes physically impossible to control and make things with. What happens if mold tries to push or pull the block that it is on?

**Comment by calebstenberg** [comment]
### Implementation Idea:
I think that whatever system we go with should be implemented with overarching graph networks. Each network should know how different "Measurement Device" nodes connect to logic nodes and output nodes. This will let us make this extremely fast by using propagated network connections instead of propagated block updates, and all connections in a network are saved in memory so we don't need to dynamically pull chunks either.
This system is easier with wireless transmission but could definitely be implemented by Mojang for redstone or a redstone/slime/string system.
### Compiled "Language"
As blocks are placed we look for networks that overlap that chunk and update the network on block-place rather than at "runtime".
This basically turns it from a interpreted "language" to a compiled "language". 
### Gates:
Redstone has emergent gates built from multiple blocks, if we analyze the input/output states of a subset of the network then we can "compile" emergent gates into very fast Boolean operations.
Or we could just have blocks for gates, but those can be hard to implement and use when it is only a 1x1 cube that needs multiple inputs and an output.


### Ores:
After the next Orevolution we could put some of this as an ore. Wait, no. Ores are per block not per block face. Maybe there is still a special "ore" that allows for signals to travel through a block.
Unless the Orevolution includes the option for per face ores it makes more sense to use a block with a special rotation mode for data transmission.

**Comment by IntegratedQuantum** [comment]
I want to point out two things here:
1. It is explicitly not a goal to make a ridiculously fast implementation, gameplay feel is more important, and the engine should be designed around that and not the other way around.
2. Any optimizations you can make by analyzing blocks and stuff is going to be overshadowed by the cost of sharing corresponding visual updates to the clients.

As for Orevolution, I think block entities are the better solution here, as it would make it a lot easier to express connections there and to animate e.g. the slime mold in fun ways.

**Comment by Argmaster** [comment]
I think it would be beneficial if this could function separately from a block grid, so you could hide the "wires" in the walls, instead of leaving them always exposed. Something like Rust or Create: Power Grid wires, just stylized to be a mold thing.  Mentioned system have two wire modes, one for hanging wires, one for placing wires on the walls with right angles, I would assume we want only the latter. This would make the system more flexible and faster than redstone block updates.
For me this seems to be the standard of convenient energy system, next to shaft+cog based systems present in Create mod or Vintage Story

**Comment by calebstenberg** [comment]
That's a good point Quantum.
Is there a pretty solid plan at the moment? Are you saying we will have block entities that store their connections to other block entities? It sounds like we don't really want wireless connections. I agree with Argmaster that it should effectively be a wire system that is skinned like slime mold, I mean we want most of the same features regardless of what it looks like. Maybe with some additional constraints or features that really make sense for slime mold.
Argmaster, wasn't there was also an idea put out there for a secondary grid for fluids, maybe the two systems could be merged? Unless someone really wants a block that is both water logged and implanted with wires.
I do think a block entity system where you get a gui/tool to configure connections would be very useful even if just for better connections to the adjacent blocks.

**Comment by Crepestrom** [comment]
we could make the slime mold like a zipline
we have nodes that have block entity connections that tell which things they are connected to

you could make normal nodes only connect to 2 things and have a seperate block type for connecting to more so that you dont have to store tons of empty data in every node

**Comment by Earthfall10** [comment]
> I don't think the mold itself should push blocks, rather it be an external block like pistons in m word game

If we want to stick with the biopunk theme perhaps the pistons could be another fungus or plant species that lives in the same ecosystem as the signal slime. It could be a carnivorous plant that uses the signals from the slime mold to know when a creature has walked into its trap, then snaps closed using its piston parts. A natural dungeon biome with useful loot in the form of trap components.
